yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -0,0 +1,808 @@
+    const newNullifierNode = new NullifierLeafPreimage(value, lowWitness.nextNullifier, lowWitness.nextIndex);
+    const insertionPath = await this.appendIndexedTree(treeId, lowResult.index, updateLowWitness, newNullifierNode);
+
+    // Since we are appending, we might have a new minimum nullifier leaf
+    this._updateMinInfo(
+      MerkleTreeId.NULLIFIER_TREE,
+      NullifierLeafPreimage,
+      [newNullifierNode, updateLowWitness],
+      [insertionIndex, lowResult.index],
+    );
+    return {
+      leafIndex: insertionIndex,
+      insertionPath: insertionPath,
+      newOrElementToUpdate: { update: false, element: newNullifierNode },
+      lowWitness: lowResult,
+    };
+  }
+
+  /**
+   * This appends a note hash to the note hash tree
+   * @param value - The note hash to be appended
+   * @returns The insertion result which contains the insertion path
+   */
+  async appendNoteHash(value: Fr): Promise<Fr[]> {
+    const tree = this.treeMap.get(MerkleTreeId.NOTE_HASH_TREE)!;
+    tree.appendLeaf(value);
+    // We use leafCount - 1 here because we would have just appended a leaf
+    const insertionPath = tree.getSiblingPath(tree.leafCount - 1n);
+    return insertionPath!;
+  }
+
+  /**
+   * This is wrapper around treeId to get the correct minimum leaf preimage
+   */
+  private getMinInfo<ID extends IndexedTreeId, O>(
+    treeId: ID,
+    O: { fromBuffer: (buffer: Buffer) => O },
+  ): { preimage: O; index: bigint } | undefined {
+    const start = this.indexedTreeMin.get(treeId);
+    if (start === undefined) {return undefined;}
+    const [buffer, index] = start;
+    return { preimage: O.fromBuffer(buffer), index };
+  }
+
+  /**
+   * This is wrapper around treeId to set the correct minimum leaf preimage
+   */
+  private setMinInfo<ID extends IndexedTreeId>(treeId: ID, preimage: { toBuffer: () => Buffer }, index: bigint): void {
+    this.indexedTreeMin.set(treeId, [preimage.toBuffer(), index]);
+  }
+
+  /**
+   * This is wrapper around treeId to set values in the indexedUpdates map
+   */
+  private setIndexedUpdates<ID extends IndexedTreeId>(treeId: ID, index: bigint, O: { toBuffer: () => Buffer }): void {
+    let updates = this.indexedUpdates.get(treeId);
+    if (updates === undefined) {
+      updates = new Map();
+      this.indexedUpdates.set(treeId, updates);
+    }
+    updates.set(index, O.toBuffer());
+  }
+
+  /**
+   * This is wrapper around treeId to get values in the indexedUpdates map
+   */
+  private getIndexedUpdates<ID extends IndexedTreeId, O>(
+    treeId: ID,
+    index: bigint,
+    O: { fromBuffer: (buffer: Buffer) => O },
+  ): O {
+    const updates = this.indexedUpdates.get(treeId);
+    if (updates === undefined) {
+      throw new Error('No updates found');
+    }
+    const buffer = updates.get(index);
+    if (buffer === undefined) {
+      throw new Error('No updates found');
+    }
+    return O.fromBuffer(buffer);
+  }
```

Comment:
Like I mentioned earlier, I really don't think we need to be doing to/fromBuffer stuff here. We should just just be able to store preimages. And we know that we can cast them to NullifierTreeLeafPreimage, PublicDataTreeLeafPreimage, etc based on the tree index